cmd/chisel/db.go

```diff
@@ -0,0 +1,250 @@
+				Path:   entry.Path,
+				Mode:   mode,
+				Slices: sliceNames,
+				Hash:   entry.Hash,
+				Size:   uint64(entry.Size),
+				Link:   entry.Link,
+			})
+			if err != nil {
+				return nil, err
+			}
+		}
+		// Add the DB path and content entries.
+		for _, pathEntry := range dbPathEntries {
+			err := dbw.Add(pathEntry)
+			if err != nil {
+				return nil, err
+			}
+			for _, s := range pathEntry.Slices {
+				err := dbw.Add(&Content{
+					Kind:  "content",
+					Slice: s,
+					Path:  pathEntry.Path,
+				})
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+
+		dbPath := dbPaths[path]
+		logf("Generating Chisel DB at %s...", dbPath)
+		err := WriteDB(dbw, dbPath, dbMode)
+		if err != nil {
+			return nil, err
+		}
+		relPath := filepath.Clean("/" + strings.TrimPrefix(dbPath, opts.RootDir))
+		generatedPaths = append(generatedPaths, relPath)
+	}
+
+	return generatedPaths, nil
+}
+
+type Package struct {
+	Kind    string `json:"kind"`
+	Name    string `json:"name"`
+	Version string `json:"version"`
+	Digest  string `json:"sha256"`
+	Arch    string `json:"arch"`
+}
+
+type Slice struct {
+	Kind string `json:"kind"`
+	Name string `json:"name"`
+}
+
+type Path struct {
+	Kind      string   `json:"kind"`
+	Path      string   `json:"path"`
+	Mode      string   `json:"mode"`
+	Slices    []string `json:"slices"`
+	Hash      string   `json:"sha256,omitempty"`
+	FinalHash string   `json:"final_sha256,omitempty"`
+	Size      uint64   `json:"size,omitempty"`
+	Link      string   `json:"link,omitempty"`
+}
+
+type Content struct {
+	Kind  string `json:"kind"`
+	Slice string `json:"slice"`
+	Path  string `json:"path"`
+}
+
+// WriteDB writes all added entries to the Chisel DB and generates the actual
+// file. It returns the path of the generated Chisel DB file. The file
+// chisel.db is a zstd compressed file.
+func WriteDB(writer *jsonwall.DBWriter, path string, mode fs.FileMode) (err error) {
+	if err = os.MkdirAll(filepath.Dir(path), 0755); err != nil {
+		return err
+	}
+
+	debugf("writing DB at %s...", path)
+	file, err := os.OpenFile(path, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, mode)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	w, err := zstd.NewWriter(file)
+	if err != nil {
+		return err
+	}
+	defer w.Close()
+
+	_, err = writer.WriteTo(w)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// locateManifests returns a map of slices which contains paths with
+// "generate:manifest". Those paths are used as keys of the returning map.
+func locateManifests(slices []*setup.Slice) map[string][]*setup.Slice {
+	manifestInfo := make(map[string][]*setup.Slice)
+	for _, s := range slices {
+		for path, info := range s.Contents {
+			if info.Generate == setup.GenerateManifest {
+				if manifestInfo[path] == nil {
+					manifestInfo[path] = []*setup.Slice{}
+				}
+				if len(manifestInfo[path]) == 0 || manifestInfo[path][len(manifestInfo[path])-1] != s {
+					manifestInfo[path] = append(manifestInfo[path], s)
+				}
```

Comment:
How does this work? Why are we only checking the last index for equality?

Reply:
Oh right, this is to prevent adding the same slice twice. I initially thought we may have the exact same path twice in a slice contents. But later found out that this is not permissible by the yaml parser. I will remove it. Thanks!